src/hocon_schema.erl

```diff
@@ -0,0 +1,106 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(hocon_schema).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export([map/2]).
+
+map(Schema, RichMap) ->
+    Keys = apply(Schema, keys, []),
+    do_map(Keys, RichMap, []).
+
+do_map([], _RichMap, Acc) ->
+    Acc;
+do_map([{Key, SchemaFun} | More], RichMap, Acc) ->
+    Value = deep_get(Key, RichMap, value),
+    Converter = SchemaFun(converter),
```

Comment:
converter could return `undefined`.
we may need a `apply_converter` function.
e.g.
apply_converter(SchemaFun, converter, Value) ->
    case SchemaFun(converter) of
        undefined ->  Value;
        F -> F(Value)
    end.
```

Reply:
and it's the same for other kinds too if they are optional.